src/hazmat/lucas.rs

```diff
@@ -547,19 +565,41 @@ pub fn lucas_test<T: Integer>(candidate: Odd<T>, base: impl LucasBase, check: Lu
         }
     }
 
+    // Double the index again:
+    vk = vk.square() - &qk - &qk; // now `vk = V_{d * 2^s} = V_{n+1}`
+
     // Lucas-V check[^Baillie2021]: if `V_{n+1} != 2 Q`, report `n` as composite.
+    let lucas_v = vk == q.double();
     if check == LucasCheck::LucasV {
-        // Double the index again:
-        vk = vk.square() - &qk.double(); // now `vk = V_{d * 2^s} = V_{n+1}`
-
-        if vk != q.double() {
+        if !lucas_v {
             return Primality::Composite;
         } else {
             return Primality::ProbablyPrime;
         }
     }
 
-    Primality::Composite
+    // The only remaining variant at this point.
+    debug_assert!(check == LucasCheck::Bpsw21);
+
+    // In case of BPSW'21, even if the Lucas-V check is passed we have another check to apply
+    if !lucas_v {
+        return Primality::Composite;
+    }
+
+    // Euler criterion: if `Q^((n+1)/2) != Q * (Q/n) mod n`, report `n` as composite.
+    let q_jacobi = jacobi_symbol_vartime(abs_q, q_is_negative, &candidate);
+    let t = match q_jacobi {
+        // This branch is unreachable since we previously checked that either `Q = 1` or `gcd(Q, n) != 1`.
+        JacobiSymbol::Zero => zero,
```

Comment:
Maybe we can enshrine this and have `unreachable!("`Q` is either 1 or coprime to `n`, hence the Jacobi symbol is not zero"),` here?